File: cpp/202.cpp
#include <set>
#include <string>
using namespace std;

class Solution {
public:
  bool isHappy(int n) {
    set<int> saw;
    while(saw.find(n) == saw.end()){
      if(n == 1) return true;
      saw.insert(n);
      int t = 0;
      for(char c : to_string(n))
        t += (c-48) * (c-48);
      n = t;
    }
    return false;
  }
};

File: python/202.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        saw = set()
        while n not in saw:
            if n == 1:
                return True
            saw.add(n)
            n = sum(pow(ord(c)-48, 2) for c in str(n))
        return False
